python/ray/serve/_private/application_state.py

```diff
@@ -84,37 +117,94 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
 
     @property
-    def deployments(self) -> List[str]:
-        """Return all deployments name from the application"""
-        if self._deployment_params is None:
+    def target_deployments(self) -> List[str]:
+        """List of target deployment names in application."""
+        if self._target_state.deployment_infos is None:
             return []
-        return [params["name"] for params in self._deployment_params]
+        return list(self._target_state.deployment_infos.keys())
+
+    def recover_target_state_from_checkpoint(
+        self, checkpoint_data: ApplicationTargetState
+    ):
+        logger.info(f"Recovering target state for app {self._name} from checkpoint.")
+        self._set_target_state(
+            checkpoint_data.deployment_infos, checkpoint_data.deleting
+        )
+
+    def _set_target_state(
+        self,
+        deployment_infos: Optional[Dict[str, DeploymentInfo]] = None,
+        deleting: bool = False,
+    ):
+        """Set application target state.
+
+        While waiting for deploy task to finish, this should be
+            (None, False)
+        When deploy task has finished and during normal operation, this should be
+            (target_deployments, False)
+        When a request to delete the application has been received, this should be
+            ({}, True)
+        """
+
+        if deleting:
+            target_state = ApplicationTargetState(dict(), True)
+            self._update_status(ApplicationStatus.DELETING)
+        else:
+            target_state = ApplicationTargetState(deployment_infos, False)
+            self._update_status(ApplicationStatus.DEPLOYING)
+
+        # Checkpoint ahead, so that if the controller crashes before we
+        # write to the target state, the target state will be recovered
+        # after the controller recovers
+        self._save_checkpoint_func(writeahead_checkpoints={self._name: target_state})
+        # Set target state
+        self._target_state = target_state
+
+    def _delete_deployment(self, name):
+        self._endpoint_state.delete_endpoint(name)
+        self._deployment_state_manager.delete_deployment(name)
 
     def delete(self):
         """Delete the application"""
-        self._update_status(ApplicationStatus.DELETING)
+        logger.info(f"Deleting application '{self._name}'")
+        self._set_target_state(deleting=True)
 
-    def deploy(self, deployment_params: List[Dict]) -> List[str]:
-        """Deploy the application.
+    def apply_deployment_info(
+        self, deployment_name: str, deployment_info: DeploymentInfo
+    ) -> bool:
+        """
+        Deploys a deployment in the application.
+
+        Returns: Whether the deployment is being updated.
+        """
+        updating = self._deployment_state_manager.deploy(
+            deployment_name, deployment_info
+        )
+
+        if deployment_info.route_prefix is not None:
+            assert deployment_info.route_prefix.startswith("/")
```

Comment:
Should we `assert` this before calling [`self._deployment_state_manager.deploy(...)`](https://github.com/ray-project/ray/pull/34328/files#diff-31975e75fad7092f05d8d5759dfeb90d8b454fc4193e4970b55914dbb87ad968R180-R182)?